pkg/arvo/app/chat-push-hook.hoon

```diff
@@ -0,0 +1,339 @@
+    %0  [~ this(state old)]
+  ==
+::
+++  on-poke
+  |=  [=mark =vase]
+  ^-  (quip card _this)
+  |^
+  =^  cards  state
+    ?+  mark  (on-poke:def mark vase)
+        %json  (chat-action (action:dejs:store !<(json vase)))
+        %chat-action  (chat-action !<(action:store vase))
+        %chat-push-hook-action  (chat-push-hook-action !<(action:hook vase))
+    ==
+  [cards this]
+  ::
+  ++  chat-action
+    |=  act=action:store
+    ^-  (quip card _state)
+    ?>  ?=(%message -.act)
+    :_  state
+    =/  =rid  (path-to-rid path.act)
+    ?.  (~(has by providers) rid)  ~
+    ::  If this is the push-hook of a proxy, prohibit writing
+    ?>  =(our.bowl ship.rid)
+    ?.  (is-permitted bowl src.bowl path.act)  ~
+    =*  letter  letter.envelope.act
+    =?  letter
+        ?&  (team:title our.bowl src.bowl)
+            ?=(%code -.letter)
+            ?=(~ output.letter)
+        ==
+      =/  =hoon  (ream expression.letter)
+      letter(output (eval:store bowl hoon))
+    =:  author.envelope.act  src.bowl
+        when.envelope.act  now.bowl
+    ==
+    [%pass / %agent [our.bowl %chat-store] %poke %chat-action !>(act)]~
+  ::
+  ++  chat-push-hook-action
+    |=  act=action:hook
+    ^-  (quip card _state)
+    ?.  (team:title our.bowl src.bowl)
+      [~ state]
+    ?-  -.act
+        %add
+      =/  =rid  (path-to-rid path.act)
+      =/  chat-path  [%mailbox path.act]
+      =/  chat-wire  [%store path.act]
+      ?:  (~(has by providers) rid)  [~ state]
+      =:  providers  (~(put ju providers) rid our.bowl)
+          allow-history  (~(put by allow-history) rid allow-history.act)
+      ==
+      :_  state
+      [%pass chat-wire %agent [our.bowl %chat-store] %watch chat-path]~
+    ::
+        %set-proxies
+      =/  =rid  (path-to-rid path.act)
+      ?>  (~(has by providers) rid)
+      ?>  (~(all in proxies.act) |=(=ship (is-permitted bowl ship path.act)))
+      =.  providers  (~(put by providers) rid (~(put in proxies.act) our.bowl))
+      [~ state]
+    ::
+        %remove
+      =/  =rid  (path-to-rid path.act)
+      ?.  (~(has by providers) rid)  [~ state]
+      =.  providers  (~(del by providers) rid)
+      =/  backlog-paths=(list path)
+        %+  murn  ~(tap by sup.bowl)
+        |=  [duct [@p =path]]
+        ^-  (unit ^path)
+        ?:  ?=([%backlog *] path)
+          `path
+        ~
+      :_  state
+      :~  [%give %kick [%mailbox path.act]~ ~]
+          :: XX what happens if backlog-paths is empty?
+          [%give %kick backlog-paths ~]
+          [%pass [%mailbox path.act] %agent [our.bowl %chat-store] %leave ~]
+      ==
+    ==
+  --
+::
+++  on-watch
+  |=  =path
+  ^-  (quip card _this)
+  |^
+  ?+  path          (on-watch:def path)
+      [%backlog *]  [(backlog t.path) this]
+      [%mailbox *]  [(mailbox t.path) this]
+  ==
+  ::
+  ++  provider-for
+    |=  [=ship =rid]
+    ^-  ^ship
+    =/  candidates  ~(tap in (~(got by providers) rid))
+    =/  idx  (mod ship (lent `(list @p)`candidates))
```

Comment:
Galaxies are the low bits, so for example descendants of `~zod` are all `0 mod 256`.  Since most active ships are descendants of a a small number of galaxies, better to do `(mug ship)` or something.